pkg/ruler/manager.go

```diff
@@ -105,8 +106,20 @@ func (r *DefaultMultiTenantManager) SyncRuleGroups(ctx context.Context, ruleGrou
 		RemoveFederatedRuleGroups(ruleGroups)
 	}
 
-	for userID, ruleGroup := range ruleGroups {
-		r.syncRulesToManager(ctx, userID, ruleGroup)
+	// Sync the rules to disk and then update the user's Prometheus Rules Manager.
+	// Since users are different, we can sync rules in parallel.
+	users := make([]string, 0, len(ruleGroups))
+	for userID := range ruleGroups {
+		users = append(users, userID)
+	}
+	err := concurrency.ForEachJob(ctx, len(users), 10, func(ctx context.Context, idx int) error {
+		userID := users[idx]
+		r.syncRulesToManager(ctx, userID, ruleGroups[userID])
+		return nil
+	})
+	if err != nil {
+		level.Error(r.logger).Log("msg", "unable to sync rules to disk", "err", err)
```

Comment:
An error here could happen only if the context is canceled (because the callback function above `return nil`) so I would keep it simple and not log any error which would be misleading.

```suggestion
		// The only error we could get here is a context canceled.
```